epsagon/events/celery.py

```diff
@@ -0,0 +1,215 @@
+    """
+    @functools.wraps(func)
+    def _signal_wrapper(*args, **kwargs):
+        """
+        Wraps original signal wrapper with try
+        """
+        try:
+            return func(*args, **kwargs)
+        except Exception as exception:  # pylint: disable=broad-except
+            event = EVENTS.get(get_event_key(*args, **kwargs))
+            if event:
+                trace_factory.add_exception(
+                    exception,
+                    traceback.format_exc()
+                )
+
+    return _signal_wrapper
+
+
+# -----------------------
+#  Celery publish events
+# -----------------------
+
+@signal_wrapper
+def wrap_before_publish(*args, **kwargs):
+    """
+    Wraps before publish signal (task.delay or task.apply_async)
+    """
+    body = kwargs.get('body', [None])[0]
+    if body:
+        body = list(body)
+
+    app_conn = import_module('celery').current_app.connection()
+    event = CeleryEvent(
+        time.time(),
+        kwargs.get('sender', ''),
+        kwargs.get('routing_key', ''),
+        body,
+        kwargs.get('headers', ''),
+        app_conn,
+    )
+
+    EVENTS[get_event_key(*args, **kwargs)] = event
+
+
+@signal_wrapper
+def wrap_after_publish(*args, **kwargs):
+    """
+    Wraps after publish signal (task.delay or task.apply_async)
+    """
+    event_key = get_event_key(*args, **kwargs)
+    event = EVENTS.get(event_key)
+    if event:
+        event.terminate()
+        trace_factory.add_event(event)
+        EVENTS.pop(event_key)
+
+
+# ----------------------
+#  Celery runner events
+# ----------------------
+
+@signal_wrapper
+def wrap_prerun(*args, **kwargs):
+    """
+    Wraps pre-run signal of worker
+    """
+    app_conn = import_module('celery').current_app.connection()
+    trace_factory.prepare()
+    runner = CeleryRunner(
+        time.time(),
+        kwargs.get('sender').name,
+        kwargs.get('task_id', ''),
+        kwargs.get('args'),
+        kwargs.get('retval'),
+        kwargs.get('state', ''),
+        app_conn,
+    )
+
+    EVENTS[get_event_key(*args, **kwargs)] = runner
+    trace_factory.set_runner(runner)
+
+
+@signal_wrapper
+def wrap_postrun(*args, **kwargs):
+    """
+    Wraps post-run signal of worker
+    """
+    event_key = get_event_key(*args, **kwargs)
+    event = EVENTS.get(event_key)
+    if event:
+        event.terminate()
+        trace_factory.send_traces()
+        EVENTS.pop(event_key)
+
+
+@signal_wrapper
+def wrap_retry(*args, **kwargs):
+    """
+    Wraps retry signal of worker
+    """
+    event = EVENTS.get(get_event_key(*args, **kwargs))
+    if event:
+        event.set_retry(kwargs.get('request', {}).get('retries', 0))
+
+
+@signal_wrapper
+def wrap_failure(*args, **kwargs):
+    """
+    Wraps failure signal of worker
+    """
+    event = EVENTS.get(get_event_key(*args, **kwargs))
+    if event:
+        event.set_exception(
+            kwargs.get('exception', Exception),
+            traceback.format_exc()
```

Comment:
Did you try that?
I'm not sure if should work, because the exception could be already caught by celery lib

Reply:
Yes, tested. This signal is being fired when celery detects and exception.